server/routerlicious/packages/lambdas-driver/src/kafka-service/runner.ts

```diff
@@ -103,30 +106,52 @@ export class KafkaRunner implements IRunner {
 	/**
 	 * Signals to stop the service
 	 */
-	public async stop(): Promise<void> {
-		if (!this.deferred || this.stopped) {
+	public async stop(caller?: string, uncaughtException?: any): Promise<void> {
+		if (this.stopped) {
 			return;
 		}
 
 		this.stopped = true;
+		try {
+			// Stop listening for new updates
+			await this.consumer.pause();
 
-		// Stop listening for new updates
-		await this.consumer.pause();
+			// Stop the partition manager
+			await this.partitionManager?.stop();
 
-		// Stop the partition manager
-		await this.partitionManager?.stop();
+			// Dispose the factory
+			await this.factory.dispose();
 
-		// Dispose the factory
-		await this.factory.dispose();
+			// Close the underlying consumer, but setting a timeout for safety
+			await promiseTimeout(30000, this.consumer.close());
 
-		// Close the underlying consumer, but setting a timeout for safety
-		await promiseTimeout(30000, this.consumer.close());
-
-		// Mark ourselves done once the partition manager has stopped
-		this.deferred.resolve();
-		this.deferred = undefined;
-		if (!this.runnerMetric.isCompleted()) {
-			this.runnerMetric.success("Kafka runner stopped");
+			// Mark ourselves done once the partition manager has stopped
+			if (caller === "uncaughtException") {
+				this.deferred?.reject({
+					uncaughtException: serializeError(uncaughtException),
+				}); // reject the promise so that the runService exits the process with exit(1)
+			} else {
+				this.deferred?.resolve();
+			}
+			this.deferred = undefined;
+			if (!this.runnerMetric.isCompleted()) {
+				this.runnerMetric.success("Kafka runner stopped");
+			}
+		} catch (error) {
+			if (!this.runnerMetric.isCompleted()) {
+				this.runnerMetric.error("Kafka runner encountered an error during stop", error);
+			}
+			if (caller === "uncaughtException") {
+				this.deferred?.reject({
```

Comment:
What's the implication of rejecting deferred here?

Reply:
so that it is caught by the runner catch block here - https://github.com/microsoft/FluidFramework/blob/edf2f5226388bb46f88b45f4fca87eb31b19aeff/server/routerlicious/packages/services-shared/src/runner.ts#L33

and ultimately process is killed in this block because runner couldnt be stopped - https://github.com/microsoft/FluidFramework/blob/edf2f5226388bb46f88b45f4fca87eb31b19aeff/server/routerlicious/packages/services-shared/src/runner.ts#L112-L123

If we resolve instead of reject, the process will exit with exit code 0.

Reply:
wouldn't uncaughtException already get handled on this listener? -> process.on("uncaughtException"){}

I'm not sure why are we only rejecting deferred in case of 'uncaughtexception' and not generally for all exceptions that come here?

Reply:
process.on("uncaughtException"){} handler takes care of gracefully shutting down (i.e. calls runner.stop), but it does not log/complete the RunService metric. By rejecting the deferred promise here, it takes care of rejecting the runningP promise https://github.com/microsoft/FluidFramework/blob/edf2f5226388bb46f88b45f4fca87eb31b19aeff/server/routerlicious/packages/services-shared/src/runner.ts#L41 and then is logged with RunService metric + process is exited with correct exit code here https://github.com/microsoft/FluidFramework/blob/edf2f5226388bb46f88b45f4fca87eb31b19aeff/server/routerlicious/packages/services-shared/src/runner.ts#L113-L121
So tldr, by rejecting the deferred promise here, we make sure to complete the deferred promise and complete the RunService metric and also exit the process with proper exit code.

We are only rejecting deferred in case 'uncaughtException' because any other exception will already be caught in runner.start and deferred will already be rejected from there before even coming here. For example -https://github.com/microsoft/FluidFramework/blob/edf2f5226388bb46f88b45f4fca87eb31b19aeff/server/routerlicious/packages/lambdas-driver/src/kafka-service/runner.ts#L96
https://github.com/microsoft/FluidFramework/blob/edf2f5226388bb46f88b45f4fca87eb31b19aeff/server/routerlicious/packages/routerlicious-base/src/alfred/runner.ts#L206

The other scenario that reaches here is sigterm, for which we dont want to reject the deferred because we dont want to capture it in RunService failure metric (because sigterm doesnt necessarily mean an exception, it can also be triggered manually by kubectl delete pod).
If it looks confusing, I am happy to change it to -
`if sigterm, then deferred?.resolve, else deferred?.reject`. I think this will be better, let me know what you think :)

Reply:
Yeah, i think that would be better. Thanks for the explanation Shubhangi. :)